pkg/bgp/speaker/speaker.go

```diff
@@ -0,0 +1,221 @@
+
+	endpointsGetter endpointsGetter
+	logger          *bgplog.Logger
+
+	// queue holds all the events to process for the Speaker.
+	queue workqueue.Interface
+
+	lock.Mutex
+	services map[k8s.ServiceID]*slim_corev1.Service
+}
+
+// OnUpdateService notifies the Speaker of an update to a service.
+func (s *Speaker) OnUpdateService(svc *slim_corev1.Service) {
+	svcID := k8s.ParseServiceID(svc)
+
+	eps := new(metallbspr.Endpoints)
+	epsFromSvc := s.endpointsGetter.GetEndpointsOfService(svcID)
+	if epsFromSvc != nil {
+		eps = convertInternalEndpoints(epsFromSvc)
+	}
+
+	s.Lock()
+	s.services[svcID] = svc
+	s.Unlock()
+
+	s.queue.Add(epEvent{
+		id:  svcID,
+		svc: convertService(svc),
+		eps: eps,
+	})
+}
+
+// OnDeleteService notifies the Speaker of a delete of a service.
+func (s *Speaker) OnDeleteService(svc *slim_corev1.Service) {
+	svcID := k8s.ParseServiceID(svc)
+
+	s.Lock()
+	delete(s.services, svcID)
+	s.Unlock()
+
+	// Passing nil as the service will force the MetalLB speaker to withdraw
+	// the BGP announcement.
+	s.queue.Add(svcEvent{
+		id:  svcID,
+		svc: nil,
+		eps: nil,
+	})
+}
+
+// OnUpdateEndpoints notifies the Speaker of an update to the backends of a
+// service.
+func (s *Speaker) OnUpdateEndpoints(eps *slim_corev1.Endpoints) {
+	svcID := k8s.ParseEndpointsID(eps)
+
+	s.Lock()
+	defer s.Unlock()
+	if svc, ok := s.services[svcID]; ok {
+		s.queue.Add(epEvent{
+			id:  svcID,
+			svc: convertService(svc),
+			eps: convertEndpoints(eps),
+		})
+	}
+}
+
+// OnUpdateNode notifies the Speaker of an update to a node.
+func (s *Speaker) OnUpdateNode(node *slim_corev1.Node) {
+	s.queue.Add(nodeEvent(&node.Labels))
+}
+
+// RegisterSvcCache registers the K8s watcher cache with this Speaker.
+func (s *Speaker) RegisterSvcCache(cache endpointsGetter) {
+	s.endpointsGetter = cache
+}
+
+// endpointsGetter abstracts the github.com/cilium/cilium/pkg/k8s.ServiceCache
+// object. The cache holds all services and endpoints (backends) from the K8s
+// watchers.
+type endpointsGetter interface {
+	GetEndpointsOfService(svcID k8s.ServiceID) *k8s.Endpoints
+}
+
+func convertService(in *slim_corev1.Service) *metallbspr.Service {
+	if in == nil {
+		return nil
+	}
+	ing := make([]v1.LoadBalancerIngress, len(in.Status.LoadBalancer.Ingress))
+	for i := range in.Status.LoadBalancer.Ingress {
+		ing[i].IP = in.Status.LoadBalancer.Ingress[i].IP
+	}
+	return &metallbspr.Service{
+		Type:          string(in.Spec.Type),
+		TrafficPolicy: string(in.Spec.ExternalTrafficPolicy),
+		Ingress:       ing,
+	}
+}
+
+func convertInternalEndpoints(in *k8s.Endpoints) *metallbspr.Endpoints {
+	if in == nil {
+		return nil
+	}
+	out := new(metallbspr.Endpoints)
+	for ip, be := range in.Backends {
+		ep := metallbspr.Endpoint{
+			IP:       ip,
+			NodeName: &be.NodeName,
+		}
+		out.Ready = append(out.Ready, ep)
+		// if be.Ready {
+		// 	out.Ready = append(out.Ready, ep)
+		// }
+		// if !be.Ready {
+		// 	out.NotReady = append(out.NotReady, ep)
+		// }
```

Comment:
To be ready or not to be ready. That is the question or is the question why is this commented?

Reply:
I had squashed a change with those commented out, with the intent that I don't need the code anymore. But I forgot to remove the commented out code :sweat_smile: